Add ability to browse through data by most recent

hulloanson opened the pull request:
TODOs

- [x] Add submission list panel

- [x] "View all" button in panel and expand submission list to the entirety of the sidebar when clicked

- [x] Click submission to focus pin

- [x] Click pin to focus submission

- ~~[ ] Focus animation for submission list when pin is clicked~~

- ~~[ ] Focus animation for pin when submission is clicked~~

- [x] Filters should apply to list of reports (will also prevent click on filtered-out issue not zooming to pin)

- [x] put feed into its own tab

- [x] focus feed tab when user wants to open it in sidebar

hulloanson committed:
gave panels in sidebar a css card; css class name option in Component

hulloanson committed:
bootstrapped submissions component

hulloanson committed:
reused issue label in submission list

tallcoleman committed:
Changed name of map component to prevent namespace conflict

using issue-map to prevent potential namespace conflict with

tallcoleman committed:
Added event handler to show marker on issue list click

hulloanson committed:
added "view all" to latest submissions

tallcoleman committed:
Added panel nav and functionality

tallcoleman committed:
Refined styling for sidebar nav

tallcoleman committed:
Organized sidebar components

tallcoleman committed:
Added toLocaleString formatting for feed item titles

tallcoleman committed:
Use filtered dataset

Ensures that pan and zoom always works for issues listed in sidebar feed

hulloanson committed:
make tab switching work again

hulloanson committed:
Merge remote-tracking branch 'upstream/main' into 103-add-ability-to-browse-through-data-by-most-recent

hulloanson committed:
switching tabs to feed with hash nav now works

hulloanson committed:
moved routing logic to a separate HashRouter

hulloanson committed:
hint about filter when no submission in list

hulloanson committed:
focus submission based on hash url and fixed focusing submission from popup

hulloanson committed:
misc cleanup

hulloanson committed:
Merge remote-tracking branch 'upstream/main' into 103-add-ability-to-browse-through-data-by-most-recent

hulloanson commented:
@tallcoleman @sonalranjit It is ready for review.

One important change made that is not directly related to the goal of this PR: 

- Uses "hash" path (like how it's done with e.g. react-dom-router's HashRouter) to partially save UI state (e.g. useful when sharing a particular submission for example)

Therefore:

- There's a new class called HashRouter. PanelNav now instantiates it and put it into SharedState

- `PanelNav`, `Submissions` and `Map` uses HashRouter

  - `PanelNav` uses it to decide which tab to show

  - `Map` listens to param changes in hash url to zoom to a marker. Clicking the "focus in sidebar" link changes the hash url to the submission
  
  - `Submission` listens to param changes in hash url to scroll to and focus a submission.
  
Please try it out and see what you think!

Also I merged several times with `main`. Can @tallcoleman the new features in main work as expected?

tallcoleman commented:
Thanks @hulloanson I'll give it a look through (and will check to make sure it works with the commits from main as well)

tallcoleman committed:
Fixing sidebar.css

Desired behaviour restored: nav buttons float (sticky at top) and filter button is 100% width, only appears when a filter is applied

hulloanson commented:
@tallcoleman ready to merge?

hulloanson committed:
updated doc on hash routing

hulloanson commented:
@tallcoleman Hash routing doc updated in README

tallcoleman committed:
Edited styling, interaction, and readme

- Styled links to look like buttons so that interaction outcome is clearer
- Moved CSS in JS to stylesheet
- Focus in sidebar action now scrolls active report to middle of visible area to account for floating header
- Clicking bikespace logo resets current hash router path
- Converted use of logical OR for null fallback to nullish coalescing operator so that code intent is clearer
- Small readme edits and additions

tallcoleman commented on bikespace_dashboard/js/components/hash_router.js:
```diff
@@ -0,0 +1,116 @@
+
+  executeCallbacks() {
+    for (const cb of this._callbacks) {
+      setTimeout(() => cb(), 0);
```
More of a question of curiosity - what's the purpose of using `setTimeout` here? To make the execution async?

hulloanson replied:
Yes. Basically puts the call to the end of the job queue as a callback is potentially compute heavy (though it shouldn't be) and I want other UI code to execute first.

tallcoleman reviewed (approved):
Fantastic job, thank you @hulloanson ! This is such a cool feature and really pulls you in to exploring all the different issues reported on the map.
